File: ir/expr/interpolation_ir.py
from dataclasses import dataclass
from common.span import SourceSpan
from generated import _pb2
from ir.expr.conversion_ir import Conversion
from ir.expr.expr_ir import ExprIR

@dataclass
class InterpolationIR(ExprIR):
    value: ExprIR
    str: str | None
    conversion: Conversion
    format_spec: ExprIR | None
    span: SourceSpan | None

    def to_proto(self):
        return _pb2.ExprIR(interpolation=_pb2.InterpolationIR(
            value=self.value.to_proto(), str=self.str, conversion=int(self.conversion),
            format_spec=self.format_spec.to_proto() if self.format_spec is not None else None,
            span=self.span.to_proto() if self.span is not None else None,
        ))

File: rust/src/ir/expr/interpolation_ir.rs
use crate::ir::expr_ir::ExprIR;
use crate::ir::nodes::{Conversion, JoinedStrIR};
use crate::ir::span_ir::SourceSpan;

#[derive(Debug, Clone)]
pub struct InterpolationIR {
    pub value: Box<ExprIR>,
    pub str: Option<String>,
    pub conversion: Conversion,
    pub format_spec: Option<JoinedStrIR>,
    pub span: Option<SourceSpan>,
}
